### ISSUE TITLE
Loop-Idiom optimization into memcpy doesn't trigger for 16-byte struct with local temporary copy

### ISSUE THREAD

**Issue by He3lixxx** [llvm/llvm-project#56162]
Given this code ([godbolt](https://godbolt.org/z/eqorzE7zE))
```c++
#include <cstdint>
#include <cstddef>

struct Data {
  uint64_t a;
  // uint64_t b;
};

void bar(Data* __restrict data_out, Data* __restrict data_in, uint64_t count_in) {
  for(uint64_t i = 0; i < count_in; ++i) {
    Data value = data_in[i];
    *data_out++ = value;
  }
}
```

clang optimizes the loop into a `memcpy` call. However, when uncommenting the second `uint64_t` member `b`, it doesn't do that anymore.

I've played around a bit with different struct sizes using
```c++
struct Data {
  char data[SIZE];
};
```
and it seems the optimization is only applied for sizes 1, 2, 4 and 8, but not for higher powers of two (16, 32, 64) and not for sizes in-between.

When changing the type of `value` to `Data&` instead of `Data`, the loop is always optimized into a `memcpy` call, regardless of `sizeof(Data)` ([godbolt](https://godbolt.org/z/dTM1xeYMq)).

**Comment by vfdff** [comment]
With my debug, I found it affected by checking **NewAI.getAllocatedType()->isSingleValueType()**, and it is brought by the first initial MR.
```
    // If this doesn't map cleanly onto the alloca type, and that type isn't
    // a single value type, just emit a memcpy.
    bool EmitMemCpy =
        !VecTy && !IntTy &&
        (BeginOffset > NewAllocaBeginOffset || EndOffset < NewAllocaEndOffset ||
         SliceSize !=
             DL.getTypeStoreSize(NewAI.getAllocatedType()).getFixedSize() ||
         !NewAI.getAllocatedType()->isSingleValueType());
```
